File: src/main/java/Chapter02/expression/ExpressionMain.java
package Chapter02.expression;

public class ExpressionMain {

    public static void main(String[] args) {
        System.out.println(
                eval(new SumJava(new SumJava(new NumJava(1), new NumJava(2)), new NumJava(4))));
    }

    public static int eval(ExpressionJava expression) {
        if (expression instanceof NumJava) {
            NumJava num = (NumJava) expression;
            return num.getValue();
        }
        if (expression instanceof SumJava) {
            return eval(((SumJava) expression).getRight()) + eval(((SumJava) expression).getLeft());
        }
        throw new IllegalArgumentException("Unknown Expression");
    }

}

File: src/main/kotlin/Chapter02/expression/ExpressionMain.kt
package Chapter02.expression

fun main() {
    println(eval(Sum(Sum(Num(1), Num(2)), Num(4))))
}

fun eval(expression: Expression): Int =
    when (expression) {
        is Num -> expression.value
        is Sum -> eval(expression.right) + eval(expression.left)
        else -> throw IllegalArgumentException("Unknown expression")
    }